### PR TITLE
Add TaskListener

### PR DESCRIPTION
**Pull request by mbasmanova**
Add a mechanism to register a listener to receive task completion events. This
is useful for logging and debugging.

### CODE REVIEW

File: velox/exec/Task.cpp [hunk 1]
```diff
@@ -22,12 +25,40 @@
...
+folly::Synchronized<std::vector<std::shared_ptr<TaskListener>>>& listeners() {
+  static folly::Synchronized<std::vector<std::shared_ptr<TaskListener>>>
+      kListeners;
+  return kListeners;
+}
+} // namespace
+
+void registerTaskListener(std::shared_ptr<TaskListener> listener) {
+  listeners().withWLock(
+      [&](auto& listeners) { listeners.push_back(std::move(listener)); });
```
Comment (majetideepak) [inline velox/exec/Task.cpp:45]: should we check if the listener is already added? It could happen due to an incorrect user application.

**Review by majetideepak — COMMENTED** [review]
@mbasmanova  great addition! Just have some minor comments.

File: velox/exec/Task.cpp [hunk 2]
```diff
@@ -22,12 +25,40 @@
...
+}
+
+bool unregisterTaskListener(const std::shared_ptr<TaskListener>& listener) {
+  return listeners().withWLock([&](auto& listeners) {
+    for (auto it = listeners.begin(); it != listeners.end(); ++it) {
```
[the remark this answers is not in the document]
Comment (majetideepak) [reply velox/exec/Task.cpp:50]: I realized `erase` needs an iterator.
I saw another erase-remove-idiom `vec.erase(std::remove(vec.begin(), vec.end(), 8), vec.end());`  here https://stackoverflow.com/questions/3385229/c-erase-vector-element-by-value-rather-than-by-position

[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/exec/Task.cpp:50]: I don't find this one more readable. What do you think?

File: velox/exec/Task.cpp (hunk 1 above)
Comment (mbasmanova) [reply velox/exec/Task.cpp:45]: We could have a check and not add a duplicate if already exists.

File: velox/exec/Task.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (majetideepak) [reply velox/exec/Task.cpp:50]: Yeah, it is verbose. The current pattern is fine.

**Comment by mbasmanova** [comment]
@majetideepak Deepak, thank you for review. Updated to address the comments.

**Review by majetideepak — APPROVED** [review]
@mbasmanova thanks, Masha.

**Review by aditi-pandit — COMMENTED** [review]
Thanks for this API.

**Pull request closed without merging**